### PR TITLE
Add tutorial for evaluating LangGraph agents

### LINKED ISSUE #1635: LangGraph Agent Evaluation Tutorial
A tutorial on how to evaluate LangGraph agents. This will provide step-by-step instructions for testing and measuring the performance of LangGraph agents.

### PR DESCRIPTION
**Pull request by sahusiddharth**
- Fixes  #1635 

This PR adds a detailed tutorial to guide users through building a ReAct agent using LangGraph. The tutorial also walks users through setting up an evaluation pipeline using Ragas to assess the agent's performance.

### CODE REVIEW

File: docs/getstarted/agent_evaluation.md [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+            if not tool_calls:
+                ragas_trace.append(r.AIMessage(content=message.content))
+            else:
+                ragas_trace.append(
+                    r.AIMessage(content=message.content, tool_calls=tool_calls)
+                )
+        elif isinstance(message, ToolMessage):
+            ragas_trace.append(r.ToolMessage(content=message.content))
+
+    return ragas_trace
```
Comment (jjmachan) [inline docs/getstarted/agent_evaluation.md:416]: can you add this as part of ragas.integrations.langchain ?

**Review by jjmachan — COMMENTED** [review]
Overall I think the tutorial is really good but I don't think this should be a getstarted blog - it's too specific to langgraph. 

What I'm thinking is add a couple of these integration blogs to 
1. langgraph
2. crewAI
3. OpenAI's swarm

ideally for the same usecase and then we'll have enough experience to build a lighter getting started tutorial that is framework agnostic

For now I think you should add this into integration/langgraph

what do you guys think?

**Comment by sahusiddharth** [comment]
Yeah, that makes sense! Adding it to the integration/langgraph section for now sounds like a good to me.

File: docs/getstarted/agent_evaluation.md (hunk 1 above)
Comment (sahusiddharth) [reply docs/getstarted/agent_evaluation.md:416]: Would it be possible to add this to langgraph instead? It seems to make more sense there since it's focused on the evaluation of agents. What do you think?

Comment (jjmachan) [reply docs/getstarted/agent_evaluation.md:416]: it depends on where the messages are defined in langgraph. If they import it from langchain then that is the root source, so we put our integration there. vice versa otherwise 

but ur call

File: docs/getstarted/agent_evaluation.jpg
Comment (shahules786) [inline docs/getstarted/agent_evaluation.jpg:1]: would be good to move this to docs/_static/imgs

File: docs/getstarted/agent_evaluation.md
Comment (shahules786) [inline docs/getstarted/agent_evaluation.md:1]: Would be good to move the document to integrations as mentioned by @jjmachan  and add a link to the document to docs/howto/integration/index.md 
We haven't populated the integration guide since the 0.2 release ( https://docs.ragas.io/en/stable/howtos/integrations/) , this can be a good next issue to take up.

**Review by shahules786 — CHANGES_REQUESTED** [review]

File: docs/getstarted/agent_evaluation.md (hunk 1 above)
Comment (sahusiddharth) [reply docs/getstarted/agent_evaluation.md:416]: Going with ragas.integrations.langgraph.

I was thinking of using the version from the Ragas library (the one we talked about above) but keeping the current version too, so people can still take inspiration from it if they want to adapt it to their own use case.

File: docs/getstarted/agent_evaluation.jpg
Comment (sahusiddharth) [reply docs/getstarted/agent_evaluation.jpg:1]: done

File: docs/getstarted/agent_evaluation.md (hunk 1 above)
Comment (sahusiddharth) [reply docs/getstarted/agent_evaluation.md:416]: Do you think I should update the tutorial as well? I was thinking of using the version from the Ragas library but keeping the `convert_to_ragas_trace` as well , so people can still take inspiration from it if they want to adapt it to their own use case. What do you think?

Comment (jjmachan) [reply docs/getstarted/agent_evaluation.md:416]: why did you end up choosing langgraph?

do push the latest changes too 🙂

File: src/ragas/integrations/langgraph.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import json
+from typing import List, Union
+
+from langchain_core.messages import AIMessage, HumanMessage, SystemMessage, ToolMessage
+
+import ragas.messages as r
+
+
+def convert_to_ragas_messages(
```
Comment (shahules786) [inline src/ragas/integrations/langgraph.py:9]: Shouldn't we have a test for this function? @jjmachan @sahusiddharth

File: docs/getstarted/agent_evaluation.md (hunk 1 above)
Comment (sahusiddharth) [reply docs/getstarted/agent_evaluation.md]: LangGraph can use components from LangChain to handle messages, but that doesn't mean that changes specific to LangGraph should be added to LangChain. Just because LangGraph achieves certain functions by using LangChain doesn't mean that LangChain itself should be responsible for changes that are only relevant to LangGraph.

Comment (jjmachan) [reply docs/getstarted/agent_evaluation.md]: fair enough, making as resolved 🙂

File: src/ragas/integrations/langgraph.py (hunk 2 above)
Comment (jjmachan) [reply src/ragas/integrations/langgraph.py:9]: yep - that would be great

Comment (sahusiddharth) [reply src/ragas/integrations/langgraph.py:9]: On it, making test cases.

I’m unsure whether I should place them in the unit tests (since it’s a single function) or in e2e tests (since this function currently represents the entire model).

File: docs/howtos/integrations/_langgraph_agent_evaluation.md [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+```python
+from ragas.integrations.langgraph import convert_to_ragas_messages
+
+# Assuming 'result["messages"]' contains the list of LangChain messages
+ragas_trace = convert_to_ragas_messages(result["messages"])
+```
+
+#### Option 2: Write a Custom Conversion Function
+Alternatively, you can implement your own conversion function by drawing inspiration from the code below. This function converts a list of LangChain messages (HumanMessage, AIMessage, ToolMessage) into Ragas-compatible message objects.
+
```
Comment (jjmachan) [inline docs/howtos/integrations/_langgraph_agent_evaluation.md:379]: why are we giving 2 options here? I think it's better to have them import it and make sure you add it in the Reference page and link it in this section. So users can check it out if they need to

File: docs/howtos/integrations/_langgraph_agent_evaluation.md [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+AI agents are becoming increasingly valuable in domains like finance, e-commerce, and customer support. These agents can autonomously interact with APIs, retrieve real-time data, and perform tasks that align with user goals. Evaluating these agents is crucial to ensure they are effective, accurate, and responsive to different inputs.
+
+In this tutorial, we'll:
+
+1. Build a [ReAct agent](https://arxiv.org/abs/2210.03629) to fetch metal prices.
+2. Set up an evaluation pipeline to track key performance metrics.
+3. Run and assess the agent's effectiveness with different queries.
+
+Click the [link](https://colab.research.google.com/drive/1IsmUfCVD_pVQ7HGsKj97BXOZBGXTIF8m?usp=sharing) to open the notebook in Google Colab.
```
Comment (jjmachan) [inline docs/howtos/integrations/_langgraph_agent_evaluation.md:11]: use this format for linking "https://colab.research.google.com/github/explodinggradients/ragas/blob/main/<path_to_notebook>.ipynb"

File: docs/howtos/integrations/_langgraph_agent_evaluation.md [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+
+## Prerequisites
+- Python 3.8+
+- OpenAI API key
```
Comment (jjmachan) [inline docs/howtos/integrations/_langgraph_agent_evaluation.md:15]: Note: depending on OpenAI is a bad idea. It's okay for now but we have to make sure that our integration pages are more accessible to people using different providers

this is something we are slowly working towards

File: docs/howtos/integrations/_langgraph_agent_evaluation.md [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+### Initializing the External Components: Setting Up the Metals.Dev API
+In this section, we'll walk through how to set up the [metals.dev](https://metals.dev/) API, which provides real-time prices for both precious and industrial metals. The API gives us a simple way to fetch the latest prices for metals like gold, silver, copper, and more.
+
+We selected metals.dev due to its free tier, which provides up to 250 requests per month — a sufficient allowance for this educational tutorials. Below are the steps to get started with the service.
+
+Before you can use the [metals.dev](https://metals.dev/) API, you will need to create an account and get an API key. Here’s how to do that:
+
+1. Go to the metals.dev website.
+2. Sign up for a free account by clicking on the Sign Up button.
+3. Once you've signed up and logged in, you will be taken to your dashboard where you can find your API key.
```
Comment (jjmachan) [inline docs/howtos/integrations/_langgraph_agent_evaluation.md:39]: I'm not sure if this is a good idea - this adds another friction point for users - now they have to sign up to another service. You had said you have a mock file right. Lets use that and just have a note pointing to the site?

Comment (jjmachan) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:39]: the goal is to make sure the users come to evaluations as fast as possible.

**Comment by jjmachan** [comment]
also could you add the integration to reference section: https://github.com/explodinggradients/ragas/blob/main/docs/references/integrations.md and provide a link to that in the docs

and add the integrations md file to https://github.com/explodinggradients/ragas/blob/895f9c02ba9c6f8c20628a215ae8b718ee3f6643/mkdocs.yml#L90-L92 so that it is part of the doc site

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 3 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:379]: Understood and,

Could you please share the process for adding it to the reference page?

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 5 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:15]: Got it

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 6 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:39]: I understand your concern, and that’s exactly why I made the account setup optional. The idea was to give users the choice: they can try the evaluation without creating an account, but for those who do, it would make their efforts much easier and more beginner-friendly.

Would you like me to remove the section about connecting to the API?

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 4 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:11]:
```
Fetch for https://api.github.com/repos/explodinggradients/ragas/contents/howtos/integrations?per_page=100&ref=docs failed: {
  "message": "No commit found for the ref docs",
  "documentation_url": "https://docs.github.com/v3/repos/contents/",
  "status": "404"
}

CustomError: Fetch for https://api.github.com/repos/explodinggradients/ragas/contents/howtos/integrations?per_page=100&ref=docs failed: {
  "message": "No commit found for the ref docs",
  "documentation_url": "https://docs.github.com/v3/repos/contents/",
  "status": "404"
}

    at new DO (https://ssl.gstatic.com/colaboratory-static/common/653a3dce72ffd5a401f2844866adbce8/external_binary_l10n__en_gb.js:2437:68)
    at va.program_ (https://ssl.gstatic.com/colaboratory-static/common/653a3dce72ffd5a401f2844866adbce8/external_binary_l10n__en_gb.js:2420:436)
    at ya (https://ssl.gstatic.com/colaboratory-static/common/653a3dce72ffd5a401f2844866adbce8/external_binary_l10n__en_gb.js:16:57)
    at va.next_ (https://ssl.gstatic.com/colaboratory-static/common/653a3dce72ffd5a401f2844866adbce8/external_binary_l10n__en_gb.js:14:195)
    at kaa.next (https://ssl.gstatic.com/colaboratory-static/common/653a3dce72ffd5a401f2844866adbce8/external_binary_l10n__en_gb.js:17:48)
    at b (https://ssl.gstatic.com/colaboratory-static/common/653a3dce72ffd5a401f2844866adbce8/external_binary_l10n__en_gb.js:17:309)

```

I’ve tried several different combinations of the path, but I’m still having trouble getting the notebook to connect properly. Could you provide more details or clarify the correct format for the path?

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 3 above)
Comment (jjmachan) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:379]: refer to this file https://github.com/explodinggradients/ragas/blob/main/docs/references/integrations.md

and you can link to it like so [convert_to_ragas_message][ragas.integrations.langgraph.convert_to_ragas_message] 

does that help?

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 6 above)
Comment (jjmachan) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:39]: yep do remove it - we want to help users spend more time on eval side of things

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 3 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:379]: Thank you for the clarification! Let me take a look at the file and review the details. I’ll get back to you shortly with any further questions or updates.

File: docs/references/integrations.md
Comment (jjmachan) [inline docs/references/integrations.md:1]: why is this deleted?

Comment (sahusiddharth) [reply docs/references/integrations.md:1]: I’m really sorry for the oversight. 

I missed adding it after making some changes. I’ll make sure to add it right away.

Comment (jjmachan) [reply docs/references/integrations.md:1]: no worries 🙂

File: docs/howtos/integrations/_langgraph_agent_evaluation.md [hunk 7]
```diff
@@ -275,7 +275,7 @@ Ragas uses its own format to evaluate agent interactions. So, if you're using La
 
 **Goal:**  Convert the list of LangChain messages (e.g., HumanMessage, AIMessage, and ToolMessage) into the format expected by Ragas, so the evaluation framework can understand and process them properly.
 
-To convert a list of LangChain messages into a format suitable for Ragas evaluation, Ragas provides the function convert_to_ragas_messages, which can be used to transform LangChain messages into the format expected by Ragas.
+To convert a list of LangChain messages into a format suitable for Ragas evaluation, Ragas provides the function [convert_to_ragas_messages](ragas.integrations.langgraph), which can be used to transform LangChain messages into the format expected by Ragas.
```
Comment (jjmachan) [inline docs/howtos/integrations/_langgraph_agent_evaluation.md:278]:
```suggestion
To convert a list of LangChain messages into a format suitable for Ragas evaluation, Ragas provides the function [convert_to_ragas_messages][ragas.integrations.langgraph.convert_to_ragas_messages], which can be used to transform LangChain messages into the format expected by Ragas.
```

need to be exact path to method (I think - will check in the rendered docs) and surrounded by square brackets

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 3 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md]: I attempted to use the function [convert_to_ragas_message](ragas.integrations.langgraph.convert_to_ragas_message), but I encountered an error stating "Unable to open 'ragas.integrations.langgraph.convert_to_ragas_message'." I also tried using [convert_to_ragas_messages](ragas.integrations.langgraph), but the same issue occurred: "Unable to open 'ragas.integrations.langgraph'."

I have made some changes please have a look at them

File: docs/howtos/integrations/_langgraph_agent_evaluation.md (hunk 7 above)
Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:278]: Should I commit the changes or wait for you confirmation? similar change would be required in md cell of ipynb file as well, right?

Comment (jjmachan) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:278]: I'm looking to see if it worked - the corrent link doesn't https://ragas--1636.org.readthedocs.build/en/1636/howtos/integrations/_langgraph_agent_evaluation/#converting-messages-to-ragas-evaluation-format

Comment (jjmachan) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:278]: can you check it locally? and then make the change? I know you have to use square brackets , not sure about using the full path (I feel that would be better)

Comment (sahusiddharth) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:278]: Are you sure about the using square brackets for both because it not linking and becomes normal text. I used something like this and it is linking but sot sure if it is the right way. I also had a look in the previous commits where something like [this](https://github.com/explodinggradients/ragas/commit/4cfa6b177bc66f6050ddbefbf43b94df8832fd8b#diff-9f80f0d8beb02e6e5931d61f3640ca25cd0b0a16daf898ee4c00ab83ce57782bR80) is used.

[convert_to_ragas_messages]\(../../../src/ragas/integrations/langgraph.py\)

Comment (jjmachan) [reply docs/howtos/integrations/_langgraph_agent_evaluation.md:278]: I'm very sure about the square brackets 
image
image
you can easily verify this locally

https://github.com/explodinggradients/ragas/blob/main/docs/getstarted/rag_testset_generation.md

**Review by jjmachan — APPROVED** [review]
Looks great! 

I know this took a lot of back and forth but you did a great job with your fast responses 🙂 - looking forward to working with you more!

**Review by shahules786 — APPROVED** [review]
Nice work, @sahusiddharth! Thanks for patiently addressing the reviews and iterating swiftly ❤️
